In [41]:
import numpy as np
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score



In [8]:
df=pd.read_csv('creditcard.csv')
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [9]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [10]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [11]:
df.shape

(284807, 31)

In [12]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [13]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

<span style="font-size:20px"> 0---->> Normal Transaction</span> 

<span style="font-size:20px"> 1---->> fraud Transaction</span> 

 


# Splitting the Normal Transaction and Fraud Transaction

In [14]:
legit = df[df['Class'] == 0]
fraud = df[df['Class'] == 1]

In [15]:
legit

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284315 rows x 31 columns]

In [16]:
fraud

Time        V1        V2        V3        V4        V5        V6  \
541        406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623        472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920      4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108      6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329      7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
541    -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623     0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920    0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108   -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329    1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
541     0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623    -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920   -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108   -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329   -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[492 rows x 31 columns]

In [17]:
legit.shape

(284315, 31)

In [18]:
fraud.shape

(492, 31)

In [19]:
legit['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [20]:
fraud['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [21]:
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]


<span style="font-size:20px">Since our data set is very unbalanced so we are taking sample of Normal Transaction to make our data set balanced</span>

In [22]:
legit_sample = legit.sample(n=492)
legit_sample

Time        V1        V2        V3        V4        V5        V6  \
22729    32404.0 -2.170640 -2.151372  1.513747 -1.022984  1.784026 -2.682317   
250115  154735.0  2.064121 -0.080258 -1.079899  0.416406 -0.174049 -1.215090   
72308    54651.0 -0.005855 -2.292529  0.497292  0.656670 -1.348502  1.100225   
190330  128789.0  0.034379  0.748816  0.099673 -0.805547  0.630032 -0.546821   
144919   86492.0 -0.168489 -0.090777  1.111993 -2.560771  0.020174 -0.306429   
...          ...       ...       ...       ...       ...       ...       ...   
190745  128969.0 -6.416152 -2.843983 -3.051791  2.213895  0.173614  1.731560   
11513    19903.0  1.303928 -0.689781  0.677966 -0.347662 -1.095737 -0.151280   
81082    58777.0  1.364356 -0.595155  0.371359 -0.615819 -1.170101 -1.093548   
208838  137295.0  1.354381 -1.512845 -0.822892  0.180104 -0.913437 -0.084932   
110237   71736.0 -1.853324 -1.122892  2.981019  2.010422 -1.771001  2.970187   

              V7        V8        V9  ...       V21       V22       V23  \
22729  -1.213508  0.069631 -1.145450  ... -0.328256 -1.181721  0.540872   
250115  0.154377 -0.309297  0.608136  ... -0.295617 -0.737223  0.343824   
72308  -0.198639  0.371405  0.883784  ...  0.207551 -0.409184 -0.357911   
190330  0.847453  0.012797 -0.157539  ... -0.248853 -0.608108  0.007047   
144919  0.463763 -0.160130  1.674862  ...  0.307091  1.228872 -0.218011   
...          ...       ...       ...  ...       ...       ...       ...   
190745  3.012332 -0.086471  0.365574  ... -0.897914 -0.389873 -1.936842   
11513  -0.990828  0.162439  0.930490  ... -0.109997 -0.085964  0.054628   
81082  -0.531401 -0.186290 -0.647627  ...  0.149932  0.258574 -0.056246   
208838 -0.304664  0.122123  1.144465  ...  0.072588 -0.555463  0.152206   
110237  1.597174  0.431183  0.096719  ...  0.459052  0.889846  0.834524   

             V24       V25       V26       V27       V28  Amount  Class  
22729   0.367689 -0.340344  0.655263  0.187150  0.230436   15.45      0  
250115  0.009140 -0.310635  0.198452 -0.073484 -0.061119    1.29      0  
72308  -0.203020 -0.261756  0.894798 -0.111287  0.094385  567.62      0  
190330 -0.520134 -0.481325  0.152872  0.237872  0.080791    2.99      0  
144919  0.665056 -0.313703 -0.182539  0.031429 -0.033182   34.95      0  
...          ...       ...       ...       ...       ...     ...    ...  
190745 -1.659699  0.802538 -0.273813  0.180473  0.424163  637.12      0  
11513   0.136389  0.324567 -0.275047 -0.014814 -0.010649    5.00      0  
81082   0.374753  0.456907 -0.206679 -0.006639  0.017980   35.89      0  
208838  0.710675 -0.702508 -0.362028 -0.071609  0.014233  325.25      0  
110237 -0.307441  0.641184  0.150183 -0.235917 -0.116414  570.49      0  

[492 rows x 31 columns]

<span style="font-size:20px">Concatenating Sample data set and fraud data set as it is now balanced data set </span>

In [23]:
concat_df = pd.concat([legit_sample , fraud],axis = 0)
concat_df

Time        V1        V2        V3        V4        V5        V6  \
22729    32404.0 -2.170640 -2.151372  1.513747 -1.022984  1.784026 -2.682317   
250115  154735.0  2.064121 -0.080258 -1.079899  0.416406 -0.174049 -1.215090   
72308    54651.0 -0.005855 -2.292529  0.497292  0.656670 -1.348502  1.100225   
190330  128789.0  0.034379  0.748816  0.099673 -0.805547  0.630032 -0.546821   
144919   86492.0 -0.168489 -0.090777  1.111993 -2.560771  0.020174 -0.306429   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
22729  -1.213508  0.069631 -1.145450  ... -0.328256 -1.181721  0.540872   
250115  0.154377 -0.309297  0.608136  ... -0.295617 -0.737223  0.343824   
72308  -0.198639  0.371405  0.883784  ...  0.207551 -0.409184 -0.357911   
190330  0.847453  0.012797 -0.157539  ... -0.248853 -0.608108  0.007047   
144919  0.463763 -0.160130  1.674862  ...  0.307091  1.228872 -0.218011   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
22729   0.367689 -0.340344  0.655263  0.187150  0.230436   15.45      0  
250115  0.009140 -0.310635  0.198452 -0.073484 -0.061119    1.29      0  
72308  -0.203020 -0.261756  0.894798 -0.111287  0.094385  567.62      0  
190330 -0.520134 -0.481325  0.152872  0.237872  0.080791    2.99      0  
144919  0.665056 -0.313703 -0.182539  0.031429 -0.033182   34.95      0  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[984 rows x 31 columns]

In [24]:
 concat_df['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [25]:
concat_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93540.815041 -0.079032 -0.062831  0.116838 -0.005892  0.028033   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.026360 -0.076423  0.039554 -0.031042  ...  0.045142  0.030662   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.014446  0.018392  0.001516 -0.002366 -0.045245  0.012191  0.002851   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       89.261199  
1      122.211321  

[2 rows x 30 columns]

In [26]:
X = concat_df.drop(columns='Class' , axis =1)
Y = concat_df['Class']  

<span style="font-size:20px">X-----> Contains all the data columns except Class column</span>

<span style="font-size:20px">Y-----> Contains Labels 0 and 1</span>

In [27]:
X

Time        V1        V2        V3        V4        V5        V6  \
22729    32404.0 -2.170640 -2.151372  1.513747 -1.022984  1.784026 -2.682317   
250115  154735.0  2.064121 -0.080258 -1.079899  0.416406 -0.174049 -1.215090   
72308    54651.0 -0.005855 -2.292529  0.497292  0.656670 -1.348502  1.100225   
190330  128789.0  0.034379  0.748816  0.099673 -0.805547  0.630032 -0.546821   
144919   86492.0 -0.168489 -0.090777  1.111993 -2.560771  0.020174 -0.306429   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
22729  -1.213508  0.069631 -1.145450  ...  0.306157 -0.328256 -1.181721   
250115  0.154377 -0.309297  0.608136  ... -0.252592 -0.295617 -0.737223   
72308  -0.198639  0.371405  0.883784  ...  0.968624  0.207551 -0.409184   
190330  0.847453  0.012797 -0.157539  ... -0.060677 -0.248853 -0.608108   
144919  0.463763 -0.160130  1.674862  ...  0.045195  0.307091  1.228872   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
22729   0.540872  0.367689 -0.340344  0.655263  0.187150  0.230436   15.45  
250115  0.343824  0.009140 -0.310635  0.198452 -0.073484 -0.061119    1.29  
72308  -0.357911 -0.203020 -0.261756  0.894798 -0.111287  0.094385  567.62  
190330  0.007047 -0.520134 -0.481325  0.152872  0.237872  0.080791    2.99  
144919 -0.218011  0.665056 -0.313703 -0.182539  0.031429 -0.033182   34.95  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [28]:
Y

22729     0
250115    0
72308     0
190330    0
144919    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

<span style="font-size:20px">Training the model</span>

In [31]:
X_train ,X_test, Y_train, Y_test = train_test_split(X,Y,test_size =0.2,stratify=Y,random_state=2)

In [32]:
model = LogisticRegression()

In [33]:
model.fit(X_train , Y_train)

C:\Users\ok\.conda\New folder\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

<span style="font-size:20px">Checking the accuracy of training data</span>

In [34]:
X_train_prediction = model.predict(X_train)
Accuracy_Train = accuracy_score(X_train_prediction,Y_train)
Accuracy_Train

0.951715374841169

<span style="font-size:20px">Checking the accuracy of test data</span>

In [35]:
X_test_prediction = model.predict(X_test)
Accuracy_Test = accuracy_score(X_test_prediction , Y_test)
Accuracy_Test

0.949238578680203

# TEST INTERFACE

In [ ]:
print("\n--- Credit Card Fraud Detection Test Interface ---")
print("Enter transaction details below. Type 'quit' at any prompt to exit.")

# Get the feature names in the order the model expects them
feature_names = X.columns.tolist()

while True:
    input_data_list = []
    try:
        print("\nPlease enter the following transaction features:")
        for feature in feature_names:
            user_input = input(f"  Enter value for {feature}: ")
            if user_input.lower() == 'quit':
                raise KeyboardInterrupt # Use an exception to break out cleanly

            # Attempt to convert input to float
            input_data_list.append(float(user_input))

        # Convert list to NumPy array and reshape for prediction (1 row, 30 columns)
        input_data_array = np.array(input_data_list).reshape(1, -1)

        # Make prediction
        prediction = model.predict(input_data_array)
        prediction_proba = model.predict_proba(input_data_array) # Get probabilities

        # Display result
        print("\n--- Prediction Result ---")
        if prediction[0] == 0:
            print("  Prediction: Legitimate Transaction")
        else:
            print("  Prediction: >> Fraudulent Transaction <<")

        print(f"  Probability -> [Legitimate: {prediction_proba[0][0]:.4f}, Fraudulent: {prediction_proba[0][1]:.4f}]")
        print("-------------------------\n")

    except ValueError:
        # Handle cases where input is not a valid number
        print("\nError: Invalid input. Please enter numeric values only.")
        print("Restarting transaction input...\n")
        continue # Skip rest of the loop and start asking for input again

    except KeyboardInterrupt:
        # Handle user typing 'quit'
        print("\nExiting test interface.")
        break # Exit the while loop

    except Exception as e:
        # Catch any other unexpected errors
        print(f"\nAn unexpected error occurred: {e}")
        print("Exiting test interface.")
        break

print("--- Interface Closed ---")

# Final Objectives

1) Data Exploration

>Loaded and explored creditcard.csv.

>Checked for null values and class distribution.

>Identified severe class imbalance (fraud cases = very few).

2) Data Preprocessing

>Separated the dataset into legit and fraud transactions.

>Took a sample of legitimate transactions to balance the dataset.

>Merged the sampled data to create a balanced training dataset.

3) Feature Selection

>Selected all columns except Class as features (X).

>Set the target label (Y) as the Class column.

4) Model Training

>Split data into training and testing sets (80/20 split, stratified).

>Trained a Logistic Regression model.

5) Model Evaluation

>Calculated training accuracy which is 95 %

>Calculated testing accuracy which is 94 %

>Used accuracy_score to evaluate predictions.